storage/test/sparse_merkle_storage/db_adapter_property_test.cpp

```diff
@@ -0,0 +1,313 @@
+//  * [getData(block) == blockUpdate] - the passed updates are the same as the fetched ones after adding the block
+//  * [getParentDigest(block) == blockDigest(prevBlock)] - blockchain property
+//  * [getLastReachableBlock() == getLatestBlock() == blockUpdates.size()] - last reachable block is equal to the latest
+//  know block ID
+TEST_P(db_adapter_block_tests, blockchain_property) {
+  const auto test = [this](const std::vector<SetOfKeyValuePairs> &blockUpdates) {
+    RC_PRE(!hasEmptyValue(blockUpdates));
+
+    auto adapter = DBAdapter{GetParam()->db()};
+    RC_ASSERT(addBlocks(blockUpdates, adapter));
+
+    RC_ASSERT(adapter.getLastReachableBlock() == blockUpdates.size());
+    RC_ASSERT(adapter.getLatestBlock() == blockUpdates.size());
+
+    for (auto i = 0ul; i < blockUpdates.size(); ++i) {
+      auto block = Sliver{};
+      RC_ASSERT(adapter.getBlockById(i + 1, block).isOK());
+      RC_ASSERT(getData(block) == blockUpdates[i]);
+      if (i == 0) {
+        RC_ASSERT(equal(getParentDigest(block), zeroDigest));
+      } else {
+        auto prevBlock = Sliver{};
+        RC_ASSERT(adapter.getBlockById(i, prevBlock).isOK());
+        RC_ASSERT(equal(getParentDigest(block), blockDigest(i, prevBlock)));
+      }
+    }
+  };
+
+  ASSERT_TRUE(rc::check(test));
+}
+
+// Test that when adapter.addBlock(block, getLastReachableBlock() + 1) is called and there is a hole between the
+// blockchain and the state transfer chain, the block is added at the back of the blockchain.
+// Note: This test uses arbitrary values for parent digest and state hash as they don't have a relation to the tested
+// property.
+TEST_P(db_adapter_block_tests, add_block_at_back_equivalent_to_last_reachable) {
+  const auto test = [this](const std::vector<SetOfKeyValuePairs> &blockchainUpdates,
+                           const std::vector<SetOfKeyValuePairs> &stateTransferUpdates,
+                           const SetOfKeyValuePairs &lastReachableUpdates,
+                           const ParentDigest &parentDigest,
+                           const Hash &stateHash) {
+    RC_PRE(!hasEmptyValue(blockchainUpdates));
+    RC_PRE(!hasEmptyValue(stateTransferUpdates));
+    RC_PRE(!hasEmptyValue(lastReachableUpdates));
+
+    // Ensure there is a hole between blockchain and state transfer blocks.
+    const auto stateTransferOffset = *rc::gen::inRange(2, 10000);
+
+    auto adapter = DBAdapter{GetParam()->db()};
+
+    // First, add blockchain updates.
+    RC_ASSERT(addBlocks(blockchainUpdates, adapter));
+
+    // Then, add state transfer blocks at an offset, ensuring a hole.
+    for (auto i = 0ul; i < stateTransferUpdates.size(); ++i) {
+      const auto blockId = adapter.getLatestBlock() + stateTransferOffset + i;
+      RC_ASSERT(adapter.addBlock(create(stateTransferUpdates[i], parentDigest.data(), stateHash), blockId).isOK());
+      RC_ASSERT(adapter.getLatestBlock() == blockId);
+    }
+
+    // Ensure there is a hole.
```

Comment:
This comment doesn't seem to match the assertion.

Reply:
Agreed. I have made it clearer. Please see the comment above.